controllers/configurationpolicy_controller.go

```diff
@@ -1963,37 +1995,48 @@ func buildNameList(
 // getNamesOfKind returns an array with names of all of the resources found
 // matching the GVK specified.
 func getNamesOfKind(
-	desiredObj unstructured.Unstructured,
+	desiredObj *unstructured.Unstructured,
 	rsrc schema.GroupVersionResource,
 	namespaced bool,
 	ns string,
 	dclient dynamic.Interface,
 	complianceType string,
 	zeroValueEqualsNil bool,
 ) (kindNameList []string) {
+	var resList *unstructured.UnstructuredList
+	var err error
+
 	if namespaced {
 		res := dclient.Resource(rsrc).Namespace(ns)
 
-		resList, err := res.List(context.TODO(), metav1.ListOptions{})
-		if err != nil {
-			log.Error(err, "Could not list resources", "rsrc", rsrc, "namespaced", namespaced)
-
-			return kindNameList
-		}
+		resList, err = res.List(context.TODO(), metav1.ListOptions{})
+	} else {
+		res := dclient.Resource(rsrc)
 
-		return buildNameList(desiredObj, complianceType, resList, zeroValueEqualsNil)
+		resList, err = res.List(context.TODO(), metav1.ListOptions{})
 	}
 
-	res := dclient.Resource(rsrc)
-
-	resList, err := res.List(context.TODO(), metav1.ListOptions{})
 	if err != nil {
 		log.Error(err, "Could not list resources", "rsrc", rsrc, "namespaced", namespaced)
 
 		return kindNameList
 	}
 
-	return buildNameList(desiredObj, complianceType, resList, zeroValueEqualsNil)
+	if len(resList.Items) == 0 {
+		return kindNameList
+	}
+
+	if desiredObj != nil {
```

Comment:
This approach makes it seem a little "magical" in the sense that if I don't provide `desiredObj` I get a completely different behavior out of this function and things such as `complianceType` are ignored.

Reply:
so you are suggesting separate func?

Reply:
Some refactoring is required. We shouldn't duplicate `res.List(` when `objNames == 0` (in the calling function) since that's a real API server call and not using a cache.

Reply:
I did